### DISCUSSION TITLE
Render dates, as in neuron

### DISCUSSION THREAD

**Discussion by Terkwood** [srid/emanote#131] [Ideas]
Hello!

Just curious whether it would be possible to render the dates associated with neuron-style notes, e.g. 

```
---
date: 2021-09-15T08:43
---
```

This might only apply to the neuron-style template.

**Comment by srid** [comment]
Dates, along with other YAML metadata, can be accessed in the HTML templates. Example, 

https://github.com/srid/emanote/blob/31aa9317a60e4053aee24005cfa3d9c822d009fa/default/templates/filters/query-timeline.tpl#L9-L13

The problem is that it will spit out an error if any note *does not* have that meta property. 

I've detailed one approach to solve this here: https://github.com/snapframework/heist/issues/128 ... basically we extend the template syntax to provide "or else" case (which can be empty text, and it has to be done in the Heist library ... however, none of my issues/PRs to that library so far has received a response from the maintainers.

**Reply by srid** [reply]
> Dates, along with other YAML metadata, can be accessed in the HTML templates

*And* we can make a small modification to Emanote to add an `after-note.tpl` template hook, similar to the [before-note.tpl](https://github.com/srid/emanote/tree/master/default/templates/hooks) - so that you can do so in a few lines, rather than override a full template file.

The before-note.tpl hook is used in the website; see [here](https://github.com/srid/emanote/tree/master/docs/templates/hooks).

**Reply by srid** [reply]
@Terkwood I've added `after-note.tpl` to master branch: https://github.com/srid/emanote/commit/663112d7556262eeba654c1edbb5bed3f9697a5b

Example usage: https://github.com/srid/inspirativity.zettel.page/commit/d41c0e688dabcf386f18e5555245b3a3ccc97352

If you visit https://inspirativity.zettel.page/ you'll see "delectation, enjoyment, appreciation, relish, zest, gusto ...
satisfaction" at the bottom of all notes (but above backlinks panel).

**Reply by Terkwood** [reply]
Fantastic! 🕺🏻

**Reply by srid** [reply]
Just added `note-end.tpl` hook https://github.com/srid/emanote/commit/e0229f1b9b3b2fc49a0aab51f098a4211734bea6 (this will put the stuff *after* the backlinks section)

**Reply by Terkwood** [reply]
I am finally caught up and have everything put together in my head. Thanks again for the quick responses. This looks completely sufficient for my own use case, where all notes have neuron-style dates. 🌟

**Reply by Terkwood** [reply]
Hi, sorry -- is this available in `0.1.53.0`?  I tried making `.emanote/templates/hooks/after-note.tpl` in my notes directory, but nothing happens.  I may not understand which file I'm supposed to edit.

**Reply by Terkwood** [reply]
OK, I updated to `0.1.60.0` using nix + git repo.  I see that I need to update `templates/hooks/after-note.tpl` to change the note rendering.  Not sure that the date approach works.  Maybe my repo isn't compatible due to a single note lacking the date.  If I try to view other notes that do have dates, they don't render.

**Reply by srid** [reply]
Technically whatever you put in your `after-note.tpl` should be part of the HTML. Did you try using your browser inspector to debug and see what's in there?

I use it here: https://github.com/srid/inspirativity.zettel.page/commit/d41c0e688dabcf386f18e5555245b3a3ccc97352

**Reply by Terkwood** [reply]
That's a good idea, I'll take a look. I replicated the simple example from inspirativity successfully, it was only the date snippet which failed

**Reply by Terkwood** [reply]
When I put the simple "delectation" example together with the `<ema:note:metadata>` tag, the latter shows up in HTML but doesn't appear to get transformed into useful output.  I can see the "delectation" blurb though.

![Screenshot from 2021-09-23 12-07-32](https://user-images.githubusercontent.com/38859656/134543800-75c0ab18-e74f-46c2-8e41-b6b0247fed23.png)

after-note.tpl
```html
<div>
  <aside class="flex items-center justify-center mt-4 text-xs text-gray-300">
  delectation, enjoyment, appreciation, relish, zest, gusto ... <a href="Satisfaction">satisfaction</a>
  </aside>
  <ema:note:metadata> 
    <span class="mr-2 text-right text-gray-600"> 
      <value var="date" /> 
    </span> 
  </ema:note:metadata> 
</div>
```

example markdown file 
```md
---
date: 2021-07-01T21:54
---

# blah blah blah
```

**Reply by srid** [reply]
It's `ema:metadata` (not `ema:note:metadata`).

**Reply by srid** [reply]
The `ema:note` prefix applies to a single result in ~~the query context~~ context of listing of note links (in query results or in tag index). Whereas the `ema:` prefix is for the *current* note. I should probably rename these to make it clear enough. Should be done as part of https://github.com/srid/emanote/issues/146 before release

**Reply by Terkwood** [reply]
It works!  Thanks so much for helping me with this!

**Comment by Terkwood** [comment]
Great, thank you! I will try editing the HTML template

**Comment by Terkwood** [comment]
For anyone who finds this discussion later, this is what I wrote in `templates/hooks/after-note.tpl` in order to display the date:

```html
<ema:metadata>
  <span class="mr-2 text-right text-gray-600">
    <value var="date" />
  </span>
</ema:metadata>
```

**Comment by fiatjaf** [comment]
For anyone bothered by the error case, this is a solution that uses our friend JavaScript to remove the date error when there is no date set on the note:

```html
<div class="flex items-center justify-center mt-2">
  <ema:metadata>
    <span id="notedate" class="mr-2 text-right text-gray-600">
      <value var="date" />
    </span>
  </ema:metadata>
  <script>if (notedate.innerText.includes('error')) notedate.innerText = ''</script>
</div>
```

**Reply by fiatjaf** [reply]
The styling not great on the example above, please change.

**Reply by srid** [reply]
Good tip! Using it here: https://github.com/nixos-asia/website/commit/1e287874a0c39d5ed427971ecb576cb519a6f9bc

**Comment by fiatjaf** [comment]
Would be nice to have a template variable for a nicely-formatted date instead of just the raw "date" string as it is in the frontmatter.

**Comment by TristanCacqueray** [comment]
To avoid the decoding error, you can also add an empty date to the index.yaml :)
